ergo/distributions/histogram.py

```diff
@@ -122,56 +119,79 @@ def rv(self):
     def destructure(self):
         return (
             HistogramDist,
-            (
-                self.logps,
-                self.ps,
-                self.cum_ps,
-                self.bins,
-                self.size,
-                self.scale_min,
-                self.scale_max,
-            ),
+            (self.logps, self.ps, self.cum_ps, self.size,),
+            *self.scale.destructure(),
         )
 
     @classmethod
-    def structure(cls, params):
+    def structure(cls, *params):
+        print(f"direct init bins are:\n {params[3]}")
         return cls(
             direct_init={
                 "logps": params[0],
                 "ps": params[1],
                 "cum_ps": params[2],
-                "bins": params[3],
-                "size": params[4],
-                "scale_min": params[5],
-                "scale_max": params[6],
+                "size": params[3],
+                "scale": params[4](*params[5]),
             }
         )
 
     @classmethod
-    def from_pairs(cls, pairs):
+    def from_pairs(cls, pairs, scale: Scale, normalized=False):
         sorted_pairs = sorted([(v["x"], v["density"]) for v in pairs])
         xs = [x for (x, density) in sorted_pairs]
+        if not normalized:
+            xs = scale.normalize_points(xs)
         densities = [density for (x, density) in sorted_pairs]
-        scale_min = xs[0]
-        scale_max = xs[-1]
-        logps = onp.log(onp.array(densities) / sum(densities))
-        return cls(logps, scale_min=scale_min, scale_max=scale_max)
 
-    def to_pairs(self):
+        # interpolate ps at normalized x bins
+        bins = onp.linspace(0, 1, len(densities) + 1)
+        target_xs = (bins[:-1] + bins[1:]) / 2
+
+        if not np.isclose(xs, target_xs, rtol=1e-04).all():
+            from scipy.interpolate import interp1d  # type: ignore
+
+            f = interp1d(xs, densities)
+            densities = f(target_xs)
+        logps = onp.log(
+            onp.array(densities) / sum(densities)
+        )  # TODO investigate why scale the densities with /sum(densities?)
```

Comment:
The bins won't exactly sum to 1 because we're approximating a continuous function with discrete steps, but the resulting discrete distribution needs to.